api/src/main/java/org/openmrs/module/attachments/obs/DefaultAttachmentHandler.java

```diff
@@ -39,20 +36,13 @@ protected ComplexData readComplexData(Obs obs, ValueComplex valueComplex, String
 				complexData.getData(), valueComplex.getMimeType()).asComplexData();
 	}
 
-	protected boolean deleteComplexData(Obs obs, AttachmentComplexData complexData) {
-		// We use a temp obs whose value complex points to the file name
-		Obs tmpObs = new Obs();
-		tmpObs.setValueComplex(complexData.asComplexData().getTitle()); // Temp obs used as a safety
-		return getParent().purgeComplexData(tmpObs);
+	protected boolean deleteComplexData(Obs obs) {
+		return getParent().purgeComplexData(obs);
 	}
 
 	protected ValueComplex saveComplexData(Obs obs, AttachmentComplexData complexData) {
 		// We invoke the parent to inherit from the file saving routines.
 		obs = getParent().saveObs(obs);
-
-		File savedFile = AbstractHandler.getComplexDataFile(obs);
-		String savedFileName = savedFile.getName();
-
-		return new ValueComplex(complexData.getInstructions(), complexData.getMimeType(), savedFileName);
+		return new ValueComplex(complexData.getInstructions(), complexData.getMimeType(), obs.getValueComplex());
```

Comment:
We've changed the constructor to take the full value complex of the obs, not the title, see below.